### DISCUSSION TITLE
Seeking advice on "ParticleSet is empty" warning during 3D simulation

### DISCUSSION THREAD

**Discussion by Caiyuan6** [Parcels-code/Parcels#1715] [Q&A] [answered]
## Question
Hi, @erikvansebille. The version of Parcels I used is 3.0.2.
I am trying to run a 3D simulation using CMEMS 3D data and the `AdvectionRK4_3D` and `CheckOutOfBounds` kernel. However, at every output time step, I encounter the following warning: WARNING: ParticleSet is empty on writing as array.

I have also tried reducing the output time (e.g., changing it to 300 seconds), but the issue still persists. The warning continues to appear at all output times.  Could you kindly provide guidance on how I can further debug this issue?

Thank you very much in advance.

Best regards,
Cyuan

## Supporting code/error messages

```python
# My partial code is as follows.
# fieldset
uvfiles = sorted(glob(dataUV_path + '*_uvcur.nc'))
wfiles = sorted(glob(dataW_path + '*_wcur.nc'))
mesh_mask = mesh_path + 'coordinate_lonup.nc'

filenames = {"U": {"lon": mesh_mask, "lat": mesh_mask, "depth": wfiles[0], "data": uvfiles},
             "V": {"lon": mesh_mask, "lat": mesh_mask, "depth": wfiles[0], "data": uvfiles},
             "W": {"lon": mesh_mask, "lat": mesh_mask, "depth": wfiles[0], "data": wfiles}, }

variables = {"U": "uo",  "V": "vo",   "W": "wo",}

grid_dimensions = {"lon": "longitude", "lat": "latitude", "depth": "depth", }

dimensions = {'U': grid_dimensions, 'V': grid_dimensions,'W': grid_dimensions, }

fieldset = FieldSet.from_netcdf(filenames, variables, dimensions, 
allow_time_extrapolation = True, mesh='spherical', timestamps = timestamps)

# kernel
def CheckOutOfBounds(particle, fieldset, time):  
    if particle.state == StatusCode.ErrorOutOfBounds:
        particle.delete()

# ParticleSet
part_lons = np.array(lons)
part_lats = np.array(lats)
rls_depth = np.ones(part_lons.shape)*0.5
pset = ParticleSet.from_list(fieldset = fieldset, pclass = MyParticle, lon = part_lons,  lat = part_lats, depth = rls_depth, )

# Kernels and execution
kernels = [AdvectionRK4_3D, CheckOutOfBounds] #, 

outputdt = timedelta(hours = 1)
runtime =  timedelta(days = 6) # the total length of the run 

pset.execute(kernels, runtime = runtime, dt = 300, output_file = output_file)

# Some other information:

![field_units](https://github.com/user-attachments/assets/77a8e4db-df2c-467a-9b50-e162a06c68cd)

![plot_code](https://github.com/user-attachments/assets/f694ae94-2fad-4519-b16c-f09bd6ae3509)
![plot_result](https://github.com/user-attachments/assets/0091d1e8-0fb3-40ff-abf0-fb76208f950e)



```

```text
# Paste your error message within this block
INFO: Output files are stored in H:\JupyterNotebookData\SCS_case3\3D\output\3D_test2.zarr.
  0%|                                                                                     | 0/518400.0 [00:00<?, ?it/s]WARNING: ParticleSet is empty on writing as array at time 0
  0%|                                                                       | 300.0/518400.0 [00:00<04:53, 1762.26it/s]WARNING: ParticleSet is empty on writing as array at time 3600
  1%|▌                                                                    | 3900.0/518400.0 [00:00<00:30, 17002.92it/s]WARNING: ParticleSet is empty on writing as array at time 7200
  1%|▉                                                                    | 7500.0/518400.0 [00:00<00:21, 24050.12it/s]WARNING: ParticleSet is empty on writing as array at time 10800
  3%|█▉                                                                  | 14400.0/518400.0 [00:00<00:16, 30510.38it/s]WARNING: ParticleSet is empty on writing as array at time 14400
  3%|██▎                                                                 | 18000.0/518400.0 [00:00<00:15, 31327.94it/s]WARNING: ParticleSet is empty on writing as array at time 18000
  4%|██▊                                                                 | 21600.0/518400.0 [00:00<00:15, 32452.41it/s]WARNING: ParticleSet is empty on writing as array at time 21600
WARNING: ParticleSet is empty on writing as array at time 25200
  5%|███▎                                                                | 25500.0/518400.0 [00:00<00:16, 30601.35it/s]WARNING: ParticleSet is empty on writing as array at time 28800
  6%|████▎                                                               | 32400.0/518400.0 [00:01<00:13, 37374.72it/s]WARNING: ParticleSet is empty on writing as array at time 32400
WARNING: ParticleSet is empty on writing as array at time 36000
  7%|████▊                                                               | 36300.0/518400.0 [00:01<00:14, 34400.13it/s]WARNING: ParticleSet is empty on writing as array at time 39600
  8%|█████▋                                                              | 43200.0/518400.0 [00:01<00:12, 38924.56it/s]WARNING: ParticleSet is empty on writing as array at time 43200

```

**Comment by michaeldenes** [comment]
Hi @Cyuan-hub, likely what is happening is that all your particles are going out of bounds, and are hence deleted due to your `CheckOutOfBounds` kernel. It looks like there are some timesteps where your particles are advected without going out of bounds, maybe try plot the trajectories from your output file, something like

```python
import xarray as xr
import matplotlib.pyplot as plt

# Load the ParticleFile
ds = xr.open_zarr(output_file)

plt.plot(ds['lon'].T, ds['lat'].T)
plt.show()
```

If you see the trajectories move to the boundaries of your domain (also check through bathymetry!) then it's simply that all of your particles are being deleted!

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes. Thanks for your quick response and suggestions! I followed the advice you provided, but I encountered some other issues during the process. 
![outfile](https://github.com/user-attachments/assets/003f8ddd-84f3-4521-9c69-8915427f93d7)

![issue2](https://github.com/user-attachments/assets/90bb7f1a-57f9-4364-aa4a-ba452b7b999a)

Could you please assist me further with these problems? I would appreciate any further assistance you can offer.

**Reply by michaeldenes** [reply]
Hi @Cyuan-hub, maybe try:
`ds = xr.open_zarr(zarrName)`, as that is the filepath to your particle file.

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes.
Thanks a lot! The `ds = xr.open_zarr(zarrName)` works successfully, but I see nothing in the figure.  Do you have any idea why that might be?

![figure1](https://github.com/user-attachments/assets/1a2b2758-925e-471d-86d5-1a71f859952f)

**Reply by michaeldenes** [reply]
@Cyuan-hub  could you try
```
plt.scatter(lons, lats)
plt.show()
```

I wonder if you particles are being initialised outside of your domain, and are hence deleted on the very first timestep.

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes,
The particle release locations should be within the domain since these positions are from my 2D simulation, where the model ran correctly.
![scatter](https://github.com/user-attachments/assets/bd2e7bc1-c9a0-468e-bb4d-73543cffc8e7)

I have also checked the units in the `FieldSet` object and created a pcolormesh of the `Level[8]` depth.

![field_units](https://github.com/user-attachments/assets/55f5864a-ef4c-4d40-919e-88d2f6057a38)

![depth](https://github.com/user-attachments/assets/975393b4-5248-4ff5-a6ea-e219296707f6)

Based on these, can they infer what might be causing the warning: "ParticleSet is empty on writing as array" at every output time step?

**Reply by michaeldenes** [reply]
That's quite a strange first plot, the axis are all over the shop! Could you provide the following information too:
```
print('fset depths', fieldset.U.depth)
print('depths range', np.nanmin(rls_depth), 'to', np.nanmax(rls_depth))
print('fset lons range', np.nanmin(fieldset.U.grid.lon), 'to', np.nanmax(fieldset.U.grid.lon))
print('lons range', np.nanmin(lons), 'to', np.nanmax(lons))
print('fset lats range', np.nanmin(fieldset.U.grid.lat), 'to', np.nanmax(fieldset.U.grid.lat))
print('lats range', np.nanmin(lats), 'to', np.nanmax(lats))
```

This will help determine if the particles are initialised within the domain.

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes.
Here are the requested results:
![result1](https://github.com/user-attachments/assets/feb078b4-98a1-48df-9b06-81d561f3b222)
![result2](https://github.com/user-attachments/assets/1575fc3b-16bb-4cdb-9df0-8c821a5df1d8)
I noticed that the latitude values seem quite strange.

**Reply by michaeldenes** [reply]
Could you print out what your lons and lats are? If `np.nanmin(lons)` throws an error, I wonder what these objects are.
Both `print(lons)` and `(print(type(lons))` would be useful.

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes.
Of course! The results of `print(lons)`, `print(type(lons)`, `print(lats)`, and `print(type(lats)` are shown below:
![image](https://github.com/user-attachments/assets/5e42f514-7963-488d-8e91-08d1951b9317)

**Reply by michaeldenes** [reply]
These look like lists of strings to me, and not lists of floats. I think you may need to convert these lons and lats into floats, and then pass into parcels! Try that and then let us know how it goes!

**Reply by Caiyuan6** [reply]
> These look like lists of strings to me, and not lists of floats. I think you may need to convert these lons and lats into floats, and then pass into parcels! Try that and then let us know how it goes!

OK, I’ll try that, and I’ll let you know as soon as I make any progress.

**Reply by michaeldenes** [reply]
Testing this on my own machine with parcels versions 3.0.2, and 3.0.5, passing lists of strings seems to work, so I'm not sure if this is really the issue.

**Reply by michaeldenes** [reply]
Perhaps you can try printing the particle lon, lat, and depth in your `CheckOutOfBounds` kernel, which will tell you where your particle is when it throws the out of bounds error. Alternatively, remove that kernel and let the `pset.execute()` command fail, and see where the out of bounds error is thrown.

**Reply by Caiyuan6** [reply]
> Testing this on my own machine with parcels versions 3.0.2, and 3.0.5, passing lists of strings seems to work, so I'm not sure if this is really the issue.

Hi, @michaeldenes. 
In my 2D simulation with version 3.0.2, lists of strings also worked without any issues.
Additionally, after converting `lons` and `lats` into floats and passing them into Parcels, the `np.nanmin(lons)` was successfully printed. However, I still get the warning: "ParticleSet is empty on writing as array" at every output time step.

#### Here are the codes and their results:
![1](https://github.com/user-attachments/assets/59fe4b53-10ca-40ac-b6f3-d18e88ed9340)
![2](https://github.com/user-attachments/assets/0261afa2-5a3f-4c43-b1fb-ebb6248eefc8)
![3](https://github.com/user-attachments/assets/675c8905-302a-45b4-9378-45f11fa13f30)
![4](https://github.com/user-attachments/assets/7a0a5c6a-774b-4b91-b4a6-39394b1da9cd)
![5](https://github.com/user-attachments/assets/14258e2d-08eb-49f4-9eb9-59075ac30001)
![6](https://github.com/user-attachments/assets/b2a23df2-c69c-4663-9870-03d9becb6adc)
![7](https://github.com/user-attachments/assets/be00c7f2-90a6-42aa-a3a7-b87311258544)
 
#### Could it be an issue with how I created the mesh_mask (the coordinates)? I don’t have any specific ideas at the moment, and I need to verify this part, as I'm still not very familiar with the process of creating the coordinates. If you have any suggestions or guidance on this, I would also greatly appreciate it!

**Reply by Caiyuan6** [reply]
> Perhaps you can try printing the particle lon, lat, and depth in your `CheckOutOfBounds` kernel, which will tell you where your particle is when it throws the out of bounds error. Alternatively, remove that kernel and let the `pset.execute()` command fail, and see where the out of bounds error is thrown.

OK, I will try it.
When remove the `CheckOutOfBounds` kernel and let the `pset.execute()` command to fail, the result is as follows: 
![11](https://github.com/user-attachments/assets/9ad86b62-66fd-4c8f-a77f-d80b1303154c)

**Reply by michaeldenes** [reply]
Without seeing the mesh file or the data, it's hard to tell, but from the print statements and plots, I think it's likely fine. Your release longitudes look to be between your min and max fieldset longitudes, and same for latitudes. Your top depth layer is at 0.494025m, so releasing particles at 0.5m should be fine also...

When you plot the fieldset, why are you performing `np.fliplr()?` What does the following produce:

```python
plt.pcolormesh(fieldset.U.grid.lon, fieldset.U.grid.lat, fieldset.U.data[0,0,:,:]
plt.show()
plt.pcolormesh(fieldset.V.grid.lon, fieldset.V.grid.lat, fieldset.V.data[0,0,:,:]
plt.show()
plt.pcolormesh(fieldset.W.grid.lon, fieldset.W.grid.lat, fieldset.W.data[0,0,:,:]
plt.show()
```
Just to see what the data looks like, and what the actual lon/lat limits of the data are for each field. It's possible that we've just focussed on the U grid, but something funky is going on in the V or W. Worth a shot!

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes. Thanks so much for your help with this! Your detailed feedback and suggestions have really boosted my skills and given me more confidence in getting better at physical oceanography and coding.

When I previously plotted the `fieldset.U` without using `np.fliplr()`, I noticed that the  figure  of the `fieldset.U` appeared clearly flipped horizontally (compared to the coastline), which seemed obviously incorrect. That’s why I added `np.fliplr()` to plot the `fieldset.U.data.`
The related codes and results are as follows:

![1](https://github.com/user-attachments/assets/e6a23d1c-d13e-4ac1-ae30-289625a2e323)
![2](https://github.com/user-attachments/assets/a34b1f58-cfe7-46f5-8b01-df7617cb8dc2)
![4](https://github.com/user-attachments/assets/7e57fe8c-16c2-452c-b441-5d5aa01121d9)
![3](https://github.com/user-attachments/assets/757fa51b-cae5-4d21-94f3-2c286cfcb8b9)
![6](https://github.com/user-attachments/assets/709bbac2-6532-4ba0-ad46-db1bc98ee92e)
![5](https://github.com/user-attachments/assets/8e590a85-1529-4d0e-a86f-a46ad914a911)

It seems like something might have happened with the longitude and latitude in my U, V, and W data or coordinate data?

**Reply by michaeldenes** [reply, marked as answer]
@Cyuan-hub  no problem at all! Yep, I think in your coordinate data you have your lons and lats flipped.
Parcels 'sees' the data like it is in the unflipped cases, so it's very likely that you have just got the order of your lons and lats wrong in the mesh_mask file you created!

**Reply by Caiyuan6** [reply]
> @Cyuan-hub no problem at all! Yep, I think in your coordinate data you have your lons and lats flipped. Parcels 'sees' the data like it is in the unflipped cases, so it's very likely that you have just got the order of your lons and lats wrong in the mesh_mask file you created!

@michaeldenes, thank you very very much! 
OK. I'll recreate the mesh_mask file, and then pass it into Parcels. And I'll let you know how it goes!

**Reply by Caiyuan6** [reply]
Hi, @michaeldenes. After correcting the longitude and latitude in the coordinate data, the issue with the warning 'ParticleSet is empty on writing as array at every output time step' has been resolved. Thanks again for your help!
